File: java/SmallestStringStartingFromLeaf988.java
import java.util.Objects;

/*
You are given the root of a binary tree where each node has a value in the range [0, 25] representing the letters 'a' to 'z'.
Return the lexicographically smallest string that starts at a leaf of this tree and ends at the root.
As a reminder, any shorter prefix of a string is lexicographically smaller.
For example, "ab" is lexicographically smaller than "aba".
A leaf of a node is a node that has no children.
*/
public class SmallestStringStartingFromLeaf988 {
    public static void main(String[] args) {
        TreeNode root = new TreeNode(0);
        root.left = new TreeNode(1);
        root.right = new TreeNode(2);
        root.left.left = new TreeNode(3);
        root.left.right = new TreeNode(4);
        root.right.left = new TreeNode(3);
        root.right.right = new TreeNode(4);
        System.out.println(smallestFromLeaf(root));


        root = new TreeNode(25);
        root.left = new TreeNode(1);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(1);
        root.left.right = new TreeNode(3);
        root.right.left = new TreeNode(0);
        root.right.right = new TreeNode(2);
        System.out.println(smallestFromLeaf(root));
    }

    public static String smallestFromLeaf(TreeNode root) {
        return depthFirstSearch(root, "");
    }

    private static String depthFirstSearch(TreeNode root, String s) {
        s = (char) (root.val + 'a') + s;
        if (root.left == null && root.right == null) return s;
        else if (root.left != null && root.right != null) {
            String left = depthFirstSearch(root.left, s), right = depthFirstSearch(root.right, s);
            return left.compareTo(right) < 0 ? left : right;
        } else return depthFirstSearch(Objects.requireNonNullElseGet(root.left, () -> root.right), s);
    }

    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
}

File: python/SmallestStringStartingFromLeaf988.py
# You are given the root of a binary tree where each node has a value in the range [0, 25] representing the letters
# 'a' to 'z'. Return the lexicographically smallest string that starts at a leaf of this tree and ends at the root.
# As a reminder, any shorter prefix of a string is lexicographically smaller. For example, "ab" is lexicographically
# smaller than "aba". A leaf of a node is a node that has no children.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def smallestFromLeaf(root: Optional[TreeNode]) -> str:
    def depthFirstSearch(node, path):
        path.insert(0, chr(node.val + ord('a')))
        if not node.left and not node.right:
            result = ''.join(path)
        elif node.left and node.right:
            result = min(depthFirstSearch(node.left, path), depthFirstSearch(node.right, path))
        elif node.left:
            result = depthFirstSearch(node.left, path)
        else:
            result = depthFirstSearch(node.right, path)
        path.pop(0)
        return result

    return depthFirstSearch(root, [])


if __name__ == "__main__":
    root_outer = TreeNode(0)
    root_outer.left = TreeNode(1)
    root_outer.right = TreeNode(2)
    root_outer.left.left = TreeNode(3)
    root_outer.left.right = TreeNode(4)
    root_outer.right.left = TreeNode(3)
    root_outer.right.right = TreeNode(4)
    print(smallestFromLeaf(root_outer))  # dba
